soc/x86/apollo_lake/soc.h

```diff
@@ -27,4 +27,9 @@
 #include "soc_gpio.h"
 #endif
 
+#include <drivers/pcie/pcie.h>
+
+#define X86_SOC_EARLY_SERIAL_PCIDEV PCIE_BDF(0, 0x18, 0)
```

Comment:
Shouldn't these info provided via dts?

Reply:
Getting early configuration **out** of devicetree was an explicit design goal here.  A value in DTS needs to go through multiple levels of compilers before it gets used in code, and without a console there's no way to debug that.

This is the console driver you use to debug your DTS hacking, basically.   If we wanted design purity, it wouldn't exist at all. :)

Reply:
not sure I follow, the values in DTS are available early during build time and before we actually start building any code

> A value in DTS needs to go through multiple levels of compilers before it gets used in code

This is not correct AFAIK, DTS data is available just like Kconfig data, there is nothing that is runtime here, all data is available in header files before we start building. Maybe I am missing something here.

Reply:
We should get this from DTS:

```DT_REG_ADDR(DT_CHOSEN(zephyr_console))```

Reply:
It's not about technical capability, it's about human reasoning capability.  My ability to know that this code is doing the right thing is very seriously hampered by having to follow it all the way through  a ton of external tooling.  And this is experimentally true, given that I personally made multiple mistakes trying to get the DTS values correct (c.f. the other change in this patch that unbreaks the driver when you write "4d4b" instead of "4b4d").

Again, this is the driver you use when all the drivers don't work.  Coupling it to other driver subsystems defeats the purpose.  If you really want all systems using the same set of frameworks, we have a console driver that does that already.  And when it didn't work, I needed to use this one to find out why.

Reply:
Could we split the difference on this DTS concern? That is to day, default to a set of #defines for the addr/size or bdf/id based on what DT_REG_ADDR/DT_REG_SIZE return from DTS for "zephyr_console", but provide a blindingly obvious way to override it if it doesn't seem to be working for some reason?

In other words, in early_serial.c something along the lines of:

```
#ifndef X86_SOC_EARLY_SERIAL_PCIDEV
#define X86_SOC_EARLY_SERIAL_PCIDEV DT_REG_ADDR(zephyr_console)
```

And if it doesn't work out of the box for whatever reason set the define in soc.h to un-block debugging?

> Experimentally, devicetree errors tend to be easy to make, and without a working console impossible to diagnose

This is fair that working with DTS can be annoying. Are there changes we could make to make debugging DTS issues easier? Everything is known at build time, we could for example produce a human readable report in the build directory so that if you want to know what the DTS data is for "zephyr_console" you can just look it up.

It is definitely the case that right now, generated files like devicetree_unfixed.h are incomprehensible.

Having said all this, I won't lose sleep if we merge as-is as long as we have an understanding that doing this in this driver is a very special case, this PR retains my +1.

Reply:
> Could we split the difference on this DTS concern? That is to day, default to a set of #defines for the addr/size or bdf/id based on what DT_REG_ADDR/DT_REG_SIZE return from DTS for "zephyr_console", but provide a blindingly obvious way to override it if it doesn't seem to be working for some reason?
> 
> In other words, in early_serial.c something along the lines of:
> 
> ```
> #ifndef X86_SOC_EARLY_SERIAL_PCIDEV
> #define X86_SOC_EARLY_SERIAL_PCIDEV DT_REG_ADDR(zephyr_console)
> ```
> 
> And if it doesn't work out of the box for whatever reason set the define in soc.h to un-block debugging?


yes, that works.